### PR TITLE
bazel - test coverage

### PR DESCRIPTION
**Pull request by polina-krukovich**
**Bazel test coverage rule**

To run tests with coverage, make sure that:
- your target or its base folder is added into [.bazel-coveragerc](https://github.com/alexkarpitski/cloud-custodian/pull/26/files#diff-6bed5a1d20d2d9d8a6464249fb2085d7)
- your target or its dependencies has  `//:.bazel-coveragerc` [in data](https://github.com/alexkarpitski/cloud-custodian/pull/26/files#diff-b8d3bcb05bd672235f0f7abd610e9c62)

Examples of usage:
`bazel coverage //tests`
`bazel coverage //tests:[test_name]`
`bazel coverage //tools/c7n_gcp/tests`
`bazel coverage //tools/c7n_azure/tests_azure`
`bazel coverage //tools/c7n_kube/tests`
`bazel coverage //tools/c7n_mailer/tests`

To see the report:
1. Install **coverage.py**: `pip install coverage`
2. Open coverage folder in terminal: `cd /tmp/C7N-cov/[module.name]`
3. Use _coverage.py_ commands to see the report:
   - `coverage combine` - combines coverage data files;
   - `coverage report` - terminal report;
   - `coverage html` - html report;
4. Useful options:
   `--skip-empty` - to exclude empty files from report;
   `--include "[regular expression]"` - to include files matching the regular expression. For example, `coverage report --include "*/c7n_gcp/*"` to include files only from _c7n_gcp_ module.

See docs: [Coverage.py command line usage](https://coverage.readthedocs.io/en/coverage-5.0.4/cmd.html)

The rule uses **[coverage.py](https://coverage.readthedocs.io/en/coverage-5.0.4/)** tool to retrieve coverage data. 

Environment variable **COVERAGE_FILE** is used to specify the coverage data file path.

**.bazel-coveragerc** - coverage configuration file used to specify sources to cover. Also parallel mode is set here, since bazel runs tests in parallel. It allows to keep the data separate during measurement, later coverage files can be combined into a single **.coverage** file with `coverage combine` command.

**Coverage.py** requires test files to have permanent location, so files from sandbox cannot be used. That's why **/home/user/.cache/bazel/\_bazel\_user/hash of the workspace dir/execroot/\_\_main\_\_** directory is used. See more: [Output directory layout](https://docs.bazel.build/versions/master/output_directories.html)

### CODE REVIEW

**Comment by YanSushchynski** [comment]
Add to the .gitignore file files with pyc-extensions. I can see cached Python-files in your commit

**Comment by alexkarpitski** [comment]
```
.vscode/launch.json.sample
.vscode/settings.json.sample
.vscode/tasks.json
c7n/version.py
tools/omnissm/functions/cfn.yaml
tools/omnissm/functions/config_example.yaml
```
These changes look suspicious.

**Comment by alexkarpitski** [comment]
Need to merge the latest changes from blaze-integration-poc into your branch.
Screenshot 2020-03-04 at 15 24 10

**Comment by alexkarpitski** [comment]
`--test_env='/path/to/cloud-custodian/'`
Could you elaborate why it is required to specify test_env?

File: c7n_py_test.bzl [hunk 1]
```diff
@@ -54,3 +54,21 @@ def c7n_py_test(name, **kwargs):
     kwargs.update(main = main_name, tags = tags + ["manual"])
     py_test(name = inner_test_name, **kwargs)
     _py_test(name = name, tags = tags, test = inner_test_name, excluded_pkgs = excluded_pkgs)
+
+def _c7n_py_cov_impl(ctx):
+    my_out = ctx.outputs.executable
+    ctx.actions.write(
+        content = "cd $C7N; pytest -n auto --cov-report html --cov-append --cov tests {}; cd -".format(" ".join([f.path for f in ctx.files.srcs])),
```
Comment (alexkarpitski) [inline c7n_py_test.bzl]: 👍

**Comment by alexkarpitski** [comment]
Holding off on merging as a possibly better approach has been found.

File: c7n_py_test.bzl [hunk 2]
```diff
@@ -17,15 +17,33 @@ def _impl(ctx):
     old_runner = ctx.attr.test[DefaultInfo].files_to_run.executable
     new_runner = ctx.actions.declare_file(ctx.attr.name)
     excluded_pkgs_command = add_exclude_pkgs_command(ctx.attr.excluded_pkgs)
+    test_name = ctx.attr.name
+    test_pkg = ctx.label.package.replace("/", ".")
+    command = ""
+
+    if ctx.configuration.coverage_enabled:
+        command = (
+            "  os.system(\"mkdir \" + os.path.join(os.getenv(\"TMPDIR\"), \"C7N-cov\"))\\\n" +
```
Comment (alexkarpitski) [inline c7n_py_test.bzl]: TMPDIR and C7N-cov are used across multiple places, let's make them constants.

File: c7n_py_test.bzl [hunk 3]
```diff
@@ -17,15 +17,33 @@ def _impl(ctx):
...
+            "  os.system(\"mkdir \" + os.path.join(os.getenv(\"TMPDIR\"), \"C7N-cov\", \"%s\"))\\\n" % (test_pkg) +
```
Comment (alexkarpitski) [inline c7n_py_test.bzl]: Also instead of calling mkdir multiple times you may try -p like `mkdir -p one/two/three`.

File: .bazel-coveragerc [hunk 4]
```diff
@@ -0,0 +1,7 @@
+[run]
+source =
+    c7n
+    tools/c7n_azure/c7n_azure
+    tools/c7n_gcp/c7n_gcp
+    tools/c7n_kube/c7n_kube
```
Comment (alexkarpitski) [inline .bazel-coveragerc]: +[mailer](https://github.com/alexkarpitski/cloud-custodian/pull/23)

File: c7n_py_test.bzl [hunk 5]
```diff
@@ -17,15 +17,33 @@ def _impl(ctx):
...
+            "  while len(os.path.basename(os.getcwd())) != 32:\\\n" +
```
[the remark this answers is not in the document]
Comment (alexkarpitski) [reply c7n_py_test.bzl]: In addition to an inline comment, a high-level overview of the core logic could be added to the PR description, like what is used, what is generated, etc.

File: tests/BUILD [hunk 6]
```diff
@@ -25,7 +25,7 @@ py_library(
         "test_offhours.py",
         "test_s3.py",
     ]),
-    data = glob(["data/**/*"]) + [":placebo_data.zip"],
+    data = glob(["data/**/*"]) + [":placebo_data.zip"] + ["//:.bazel-coveragerc"],
```
Comment (alexkarpitski) [inline tests/BUILD]: I deem this should also be noted as an additional step in the main description.

Comment (alexkarpitski) [reply tests/BUILD]: Besides, is not it enough to export the file only in this BUILD? If so, please disregard the previous comment, just mention `.bazel-coveragerc` config is used in the PR description.

Comment (polina-krukovich) [reply tests/BUILD]: It's not enough. When running tests from tools/../tests and .bazel-coveragerc file is not specified in it's BUILD file, it is not added to the output folder and cannot be found.

**Pull request merged**